### ISSUE TITLE
The gradient cannot be modified under Stage3

### ISSUE THREAD

**Issue by Hashmapw** [huggingface/transformers#36348]
I use the API function `deepspeed.utils.safe_get_full_grad(param)` and `deepspeed.utils.safe_set_full_grad(param, fp32_grad)` to get and set the grad for the param, but it seemed that it doesn't work. Beacuse the result after the modified didn't have any change even i transpose the gradient with random numbers. 

I overwrite some function in class Trainer.

```python
    def compute_loss(self, model, inputs, return_outputs=False):
       [...some code...]
        if return_outputs:
            return loss, outputs
        return loss

    def training_step(self, model, inputs):
        model.train()
        loss = self.compute_loss(model, inputs)
        if self.args.gradient_accumulation_steps > 1:
            loss = loss / self.args.gradient_accumulation_steps
        self.accelerator.backward(loss)

        dataloader = self.get_train_dataloader()
        self.modulation()   # here i modified the grad

        return loss
```
